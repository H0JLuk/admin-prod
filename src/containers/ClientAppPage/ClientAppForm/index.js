import ClientAppForm from './ClientAppForm';

export default ClientAppForm;
