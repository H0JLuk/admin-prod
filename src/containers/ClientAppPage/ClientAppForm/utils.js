import { notification } from 'antd';
import { addSettings, updateSettingsList } from '../../../api/services/settingsService';
import { designKeysForCheck, SETTINGS_TYPES } from './ClientAppFormConstants';

export const showNotify = (message, isError) => {
    const config = {
        duration: 5,
        placement: 'bottomRight',
    };

    if (isError) {
        notification.error({
            ...config,
            message,
        });
    } else {
        notification.success({
            ...config,
            message,
        });
    }

};

export async function createOrUpdateKey (changedParams) {
    const updateSettings = [];
    const addSettingsArr = [];

    changedParams.forEach(({ type, ...params }) => {
        type === SETTINGS_TYPES.EDIT
            ? updateSettings.push(params)
            : addSettingsArr.push(params);
    });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}

export const checkExistDesignSettings = (settings) => {
    return designKeysForCheck.some(key => !Object.prototype.hasOwnProperty.call(settings, key));
};

export const getImage = (imageName) => {
    return `${process.env.PUBLIC_URL}/images/clientAppDesign/${imageName}.png`;
};


