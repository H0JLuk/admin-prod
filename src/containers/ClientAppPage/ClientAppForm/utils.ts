import { customNotifications } from '@utils/notifications';
import { ArgsProps } from 'antd/lib/notification';
import { addSettings, updateSettingsList } from '@apiServices/settingsService';
import { SETTINGS_TYPES } from './ClientAppFormConstants';
import { SettingDto } from '@types';
import ROLES from '@constants/roles';

export interface IChangedParam extends Pick<SettingDto, 'clientAppCode' | 'key' | 'value' | 'userRole'> {
    type: SETTINGS_TYPES;
}

export const showNotify = (message: React.ReactNode, isError?: boolean) => {
    const config: ArgsProps = {
        duration: 5,
        message,
    };

    if (isError) {
        customNotifications.error(config);
    } else {
        customNotifications.success(config);
    }

};

type UpdateAndAddSettings = {
    updateSettings: SettingDto[];
    addSettingsArr: SettingDto[];
};

export async function createOrUpdateKey(changedParams: IChangedParam[]) {
    const { updateSettings, addSettingsArr } = changedParams.reduce<UpdateAndAddSettings>((result, { type, ...params }) => {
        const key = type === SETTINGS_TYPES.EDIT ? 'updateSettings' : 'addSettingsArr';

        if (params.key === 'referralTokenLifetime') {
            delete params.clientAppCode;
            params.key = 'token_lifetime';
            params.userRole = ROLES.REFERAL_LINK;
        }

        return {
            ...result,
            [key]: [...result[key], params],
        };
    }, { updateSettings: [], addSettingsArr: [] });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}
