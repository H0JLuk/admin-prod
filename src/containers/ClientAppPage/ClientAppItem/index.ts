import ClientAppItem from './ClientAppItem';

export default ClientAppItem;
