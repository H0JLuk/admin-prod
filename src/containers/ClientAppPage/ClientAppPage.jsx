import React, { Component, Fragment } from 'react';
import { addClientApp, getClientAppList, updateClientApp } from '../../api/services/clientAppService';
import { getStaticUrl } from '../../api/services/settingsService';
import { goApp } from '../../utils/appNavigation';
import styles from './ClientAppPage.module.css';
import CustomModal from '../../components/CustomModal/CustomModal';
import ClientAppItem from '../../components/ClientAppItem/ClientAppItem';
import { getErrorText } from '../../constants/errors';
import ButtonLabels from '../../components/Button/ButtonLables';
import cross from '../../static/images/cross.svg';
import Form from '../../components/Form/Form';
import { populateFormWithData } from '../../components/Form/formHelper';
import { CLIENT_APP_ADD_FORM, CLIENT_APP_EDIT_FORM } from '../../components/Form/forms';
import { withRouter } from 'react-router-dom';
import { getRole, saveAppCode } from '../../api/services/sessionService';
import { ROUTE } from '../../constants/route';
import { logout } from '../../api/services/authService';

const CLIENT_APP_LIST_TITLE = 'Клиенские приложения';
const LOADING_LIST_LABEL = 'Загрузка';
const CLIENT_APPS_GET_ERROR = 'Ошибка получения клиентских приложений!';
const ENTER_CLIENT_APP_CODE_REQUEST = 'Пожалуйста, введите код клиентского приложения';
const ENTER_CLIENT_APP_NAME_REQUEST = 'Пожалуйста, введите имя клиентского приложения';

const initialState = {
    editingClientApp: {
        id: null, name: null, code: null, isDeleted: false
    },
    staticUrl: getStaticUrl(), clientAppList: [], isOpen: false, formError: null
};

const LoadingStatus = ({ loading }) => (
    <p className={ styles.loadingLabel }>{ loading ? LOADING_LIST_LABEL : CLIENT_APPS_GET_ERROR }</p>
);

class ClientAppPage extends Component {
    state = { ...initialState }

    componentDidMount() {
        getClientAppList().then( response => {
            const { clientApplicationDtoList } = response;
                this.setState({ clientAppList: clientApplicationDtoList });
            }).catch( () => {
                this.setState({ clientAppList: [] });
            });
    }

    clearState = () => {
        this.setState(initialState.editingClientApp);
    };

    openModal = () => {
        this.setState({ isOpen: true });
    };

    closeModal = () => {
        this.setState({ isOpen: false }, this.clearState);
    };

    handleEdit = (id, name, code, isDeleted) => {
        this.setState({ editingClientApp: { id, name, code, isDeleted } }, () => {
            this.openModal();
        });
    };

    handleAdministrate = (code) => {
        const role = getRole();
        const { history } = this.props;
        saveAppCode(code);
        goApp(history, role);
    };

    handleInputChange = (event) => {
        const target = event.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;

        this.setState({
            editingClientApp: { ...this.state.editingClientApp, isDeleted: value }
        });
    }

    doLogout = () => {
        logout().then(() => {
            this.props.history.push(ROUTE.LOGIN);
        });
    };

    renderModalForm = () => {
        const { formError, editingClientApp } = this.state;
        const formData = editingClientApp.id !== null ? populateFormWithData(CLIENT_APP_EDIT_FORM, {
            name: editingClientApp.name,
            code: editingClientApp.code,
        }) : CLIENT_APP_ADD_FORM;
        return (
            <div className={ styles.modalForm }>
                <img src={ cross }
                     onClick={ this.closeModal }
                     className={ styles.crossSvg }
                     alt={ ButtonLabels.CLOSE }
                />
                <Form
                    data={ formData }
                    buttonText={ ButtonLabels.SAVE }
                    onSubmit={ this.onSubmit }
                    formClassName={ styles.clientAppForm }
                    fieldClassName={ styles.clientAppForm__field }
                    activeLabelClassName={ styles.clientAppForm__field__activeLabel }
                    buttonClassName={ styles.clientAppForm__button }
                    errorText={ getErrorText(formError) }
                    formError={ !!formError }
                    errorClassName={ styles.error }
                />
                <form className={ styles.clientAppForm }>
                    <label className={ styles.clientAppForm__field }>
                        Is Deleted:
                    </label>
                    <input
                        name="Deleted"
                        type="checkbox"
                        checked={ editingClientApp.isDeleted }
                        onChange={ this.handleInputChange } />
                </form>
            </div>
        );
    }

    reloadClientApps = (clientAppDto) => {
        const { clientAppList, editingClientApp: { id } } = this.state;
        const newClientAppList = clientAppList.slice();
        newClientAppList.forEach( app => {
            if (app.id === id) {
                app.name = clientAppDto.name;
                app.code = clientAppDto.code;
                app.isDeleted = clientAppDto.isDeleted;
            }
        });
        this.setState({ clientAppList: newClientAppList }, this.closeModal);
    };

    pushToClientAppList(id, clientAppDto) {
        const newClientApp = {
            ...clientAppDto,
            id: id
        };
        const clientAppList = [...this.state.clientAppList, newClientApp];
        this.setState({ clientAppList }, this.closeModal);
    }

    onSubmit = (data) => {
        if (!data.code) {
            alert(ENTER_CLIENT_APP_CODE_REQUEST);
            return;
        }
        if (!data.name) {
            alert(ENTER_CLIENT_APP_NAME_REQUEST);
            return;
        }
        let clientAppDto;
        clientAppDto = {
            code: data.code,
            name: data.name,
            isDeleted: this.state.editingClientApp.isDeleted
        };
        if (this.state.editingClientApp.id !== null) {
            updateClientApp(this.state.editingClientApp.id, clientAppDto).then(() => {
                this.reloadClientApps(clientAppDto);
            }).catch(error => {
                console.log(error.message);
            });
        } else {
            addClientApp(clientAppDto).then((response) => {
                this.pushToClientAppList(response.id, clientAppDto);
            });
        }
    };

    renderClientAppList = () => {
        const { clientAppList } = this.state;
        const isSuccess = Array.isArray(clientAppList);
        return (
            <Fragment>
                {
                    (isSuccess && clientAppList.length) ?
                        clientAppList.map((app, i) =>
                            <ClientAppItem
                                key={ `clientAppItem-${i}` }
                                handleEdit={ this.handleEdit }
                                handleAdministrate={ this.handleAdministrate }
                                { ...app }
                            />) : <LoadingStatus loading />
                }
            </Fragment>
        );
    };

    renderModifyModal = () => (
        <CustomModal
            isOpen={ this.state.isOpen }
            onRequestClose={ this.closeModal }>
            {this.renderModalForm()}
        </CustomModal>
    );

    render() {
        return (
            <div className={ styles.wrapper }>
                <div className={ styles.clientAppPageWrapper }>
                    {this.renderModifyModal()}
                    <div className={ styles.headerSection }>
                        <h3>{CLIENT_APP_LIST_TITLE}</h3>
                        <div className={ styles.logout } onClick={ this.doLogout }>{ ButtonLabels.LOGOUT }</div>
                    </div>
                    <div className={ styles.clientAppList }>
                        {this.renderClientAppList()}
                    </div>
                </div>
            </div>
        );
    }
}

export default withRouter(ClientAppPage);