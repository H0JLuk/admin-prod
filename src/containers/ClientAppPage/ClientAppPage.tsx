import React, { useCallback, useEffect, useState, useMemo, useRef } from 'react';
import { SortableContainer, SortableContainerProps, SortableElement } from 'react-sortable-hoc';
import { SyncOutlined } from '@ant-design/icons';
import { History } from 'history';

import Header from '@components/Header';
import HeaderWithActions, { ButtonProps } from '@components/HeaderWithActions';
import ClientAppItem from './ClientAppItem';

import { getActiveClientApps, reorderClientApp } from '@apiServices/clientAppService';
import { getRole } from '@apiServices/sessionService';
import { arrayMove } from '@utils/helper';
import { CLIENT_APPS_PAGES } from '@constants/route';
import ROLES from '@constants/roles';
import { BUTTON_TEXT } from '@constants/common';
import { ClientAppDto } from '@types';

import style from './ClientAppPage.module.css';

const SortableContainerList = SortableContainer((props: SortableContainerListProps) => <div>{props.children}</div>);
const SortableElementItem = SortableElement(ClientAppItem);

const SEARCH_INPUT_PLACEHOLDER = 'Поиск по названию';

const DROPDOWN_SORT_MENU = [
    { name: 'displayName', label: 'По названию' },
    { name: 'code', label: 'По коду' },
];

type ClientAppPageProps = {
    matchPath: string;
    history: History;
};

type SortableContainerListProps = {
    children: Array<SortableContainerProps>;
};


const ClientAppPage: React.FC<ClientAppPageProps> = ({ matchPath, history }) => {
    const [dataList, setDataList] = useState<ClientAppDto[]>([]);
    const copyDataList = useRef<ClientAppDto[]>([]);
    const [isLoading, setIsLoading] = useState(true);
    const [isSortable, setIsSortable] = useState(false);

    const role = useMemo(() => getRole(), []);

    const loadData = useCallback(async () => {
        setIsLoading(true);

        try {
            const clientAppList = await getActiveClientApps();
            copyDataList.current = clientAppList;
            setDataList(clientAppList);
        } catch (error) {
            console.warn(error);
        }

        setIsLoading(false);
    }, []);

    useEffect(() => {
        loadData();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [loadData]);

    const toggleSortable = () => setIsSortable((state) => !state);

    const onSortEnd = useCallback(({ oldIndex, newIndex }) => {
        setDataList((state) => arrayMove(state, oldIndex, newIndex));
    }, []);

    const onCancelSortable = async () => {
        setDataList(copyDataList.current);
        setIsSortable(false);
    };

    const onSave = async () => {
        const idMap = dataList.reduce<Record<number, number>>((result, elem, index) => ({ ...result, [elem.id]: index }), {});

        setIsLoading(true);
        try {
            await reorderClientApp({ idMap });
            await loadData();
        } catch (error) {
            console.warn(error);
        }

        setIsLoading(false);
        setIsSortable(false);
    };

    const onAddClick = () => history.push(`${ matchPath }${ CLIENT_APPS_PAGES.ADD_APP }`);

    const buttons: ButtonProps[] = [];

    if (role === ROLES.ADMIN) {

        if (isSortable) {
            buttons.push(
                { type: 'primary', label: BUTTON_TEXT.SAVE, onClick: onSave },
                { label: BUTTON_TEXT.CANCEL, onClick: onCancelSortable },
            );
        } else {
            buttons.push(
                { type: 'primary', label: BUTTON_TEXT.ADD, onClick: onAddClick },
                { label: BUTTON_TEXT.CHANGE_ORDER, onClick: toggleSortable, disabled: (params) => !!(params?.sortBy || params?.filterText) },
            );
        }
    }

    return (
        <>
            {isLoading && (
                <div className={style.loadingContainer}>
                    <div className={style.loading}>
                        <SyncOutlined spin />
                    </div>
                </div>
            )}
            <div className={style.container}>
                <Header buttonBack={false} />
                <HeaderWithActions<ClientAppDto>
                    buttons={buttons}
                    setDataList={setDataList}
                    matchPath={matchPath}
                    copyDataList={copyDataList.current}
                    inputPlaceholder={SEARCH_INPUT_PLACEHOLDER}
                    filterByFieldKey="displayName"
                    menuItems={DROPDOWN_SORT_MENU}
                />
                <div className={style.content}>
                    <SortableContainerList
                        useWindowAsScrollContainer
                        onSortEnd={onSortEnd}
                    >
                        {dataList.map((item, index) => (
                            <SortableElementItem
                                key={item.id}
                                item={item}
                                isSortable={isSortable}
                                disabled={!isSortable}
                                index={index}
                                matchPath={matchPath}
                                history={history}
                                forceUpdate={loadData}
                                tooltipIsVisible={role === ROLES.ADMIN}
                            />
                        ))}
                    </SortableContainerList>
                </div>
            </div>
        </>
    );
};

export default ClientAppPage;
