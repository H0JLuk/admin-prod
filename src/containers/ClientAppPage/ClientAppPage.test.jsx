import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import ClientAppPage from './ClientAppPage';
import ROLES from '../../constants/roles';

import * as clientAppService from '../../api/services/clientAppService';
import * as sessionService from '../../api/services/sessionService';
import * as helper from '../../utils/helper';
import { CLIENT_APPS_PAGES } from '../../constants/route';
import { clientAppListTestResponse } from '../../../__tests__/constants';
import { sleep } from '../../setupTests';

describe('<ClientAppPage /> test', () => {
    const props = {
        matchPath: 'matchPath',
        history: {
            push: jest.fn(),
        },
    };

    beforeEach(() => {
        sessionService.getRole = jest.fn(() => ROLES.ADMIN);
    });

    it('should match snapshot', () => {
        const ClientAppPageComponent = shallow(<ClientAppPage { ...props } />);

        expect(ClientAppPageComponent.debug()).toMatchSnapshot();
    });

    it('should do history push on clicking add button', () => {
        const ClientAppPageComponent = shallow(
            <ClientAppPage { ...props } />
        );

        ClientAppPageComponent.find('HeaderWithActions')
            .prop('buttons')[0]
            .onClick();

        expect(props.history.push).toBeCalledTimes(1);
        expect(props.history.push).toBeCalledWith(`${props.matchPath}${CLIENT_APPS_PAGES.ADD_APP}`);
    });

    it('should go to sortable mode on clicking sort button', () => {
        clientAppService.reorderClientApp = jest.fn();
        clientAppService.getClientAppList = jest.fn(() => clientAppListTestResponse);
        const ClientAppPageComponent = shallow(<ClientAppPage { ...props } />);

        expect(ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].label).not.toEqual('Отменить');
        ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].onClick();
        expect(ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].label).toEqual('Отменить');

        ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[0].onClick();
        expect(clientAppService.reorderClientApp).toBeCalledTimes(1);
    });

    it('should warn with error in console after getClientAppList rejection', async () => {
        clientAppService.getClientAppList.mockRejectedValueOnce('warn');
        const spy = jest.spyOn(console, 'warn').mockImplementation(() => '');
        mount(
            <MemoryRouter>
                <ClientAppPage { ...props } />
            </MemoryRouter>
        );
        await sleep();
        expect(spy).toBeCalledTimes(1);
    });

    it('should cancel sortable mode', () => {
        clientAppService.reorderClientApp = jest.fn();
        clientAppService.getClientAppList = jest.fn(() => clientAppListTestResponse);
        const ClientAppPageComponent = shallow(<ClientAppPage { ...props } />);

        ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].onClick();
        expect(ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].label).toEqual('Отменить');

        ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].onClick();
        expect(ClientAppPageComponent.find('HeaderWithActions').prop('buttons')[1].label).toEqual('Изменить порядок');
        expect(1).toEqual(1);
    });

    it('should call `arrayMove` function', async () => {
        sessionService.getRole = jest.fn(() => ROLES.ADMIN);
        clientAppService.getClientAppList = jest.fn(() => clientAppListTestResponse);
        helper.arrayMove = jest.fn(() => []);

        const ClientAppPageComponent = shallow(<ClientAppPage { ...props } />);
        ClientAppPageComponent.find('sortableList').simulate('sortEnd', { oldIndex: 0, newIndex: 1 });
        expect(helper.arrayMove).toHaveBeenCalledWith([], 0, 1);
    });

    it('sortableContainer should render <div /> children', async () => {
        clientAppService.reorderClientApp = jest.fn();
        sessionService.getRole = jest.fn(() => ROLES.ADMIN);
        clientAppService.getClientAppList = jest.fn(() => clientAppListTestResponse);

        const ClientAppPageComponent = mount(
            <MemoryRouter>
                <ClientAppPage { ...props } />
            </MemoryRouter>
        );

        await act(async () => {
            expect(ClientAppPageComponent.find('sortableList').find('div')).toHaveLength(1);
        });
    });
});
