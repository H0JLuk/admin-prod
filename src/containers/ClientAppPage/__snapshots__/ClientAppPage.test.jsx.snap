// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientAppPage /> test should match snapshot 1`] = `
"<Fragment>
  <div className=\\"loadingContainer\\">
    <div className=\\"loading\\">
      <ForwardRef(SyncOutlined) spin={true} />
    </div>
  </div>
  <div className=\\"container\\">
    <Header buttonBack={false} />
    <HeaderWithActions buttons={{...}} setDataList={[Function: bound _dispatchAction]} matchPath=\\"matchPath\\" copyDataList={{...}} inputPlaceholder=\\"Поиск по названию\\" filterByFieldKey=\\"displayName\\" menuItems={{...}} />
    <div className=\\"content\\">
      <sortableList useWindowAsScrollContainer={true} onSortEnd={[Function (anonymous)]} axis=\\"y\\" disableAutoscroll={false} distance={0} getHelperDimensions={[Function: defaultGetHelperDimensions]} hideSortableGhost={true} lockOffset=\\"50%\\" lockToContainerEdges={false} pressDelay={0} pressThreshold={5} keyCodes={{...}} shouldCancelStart={[Function: defaultShouldCancelStart]} transitionDuration={300} />
    </div>
  </div>
</Fragment>"
`;
